src/libraries/System.Private.CoreLib/src/System/IO/FileSystemInfo.cs

```diff
@@ -107,6 +115,46 @@ public DateTime LastWriteTimeUtc
             set => LastWriteTimeCore = File.GetUtcDateTimeOffset(value);
         }
 
+        /// <summary>
+        /// If this <see cref="FileSystemInfo"/> instance represents a link, returns the link target's path.
+        /// If a link does not exist in <see cref="FullName"/>, or this instance does not represent a link, returns <see langword="null"/>.
+        /// </summary>
+        public string? LinkTarget => _linkTarget ??= FileSystem.GetLinkTarget(FullPath, this is DirectoryInfo);
+
```

Comment:
In https://github.com/dotnet/runtime/issues/24271 there was general consensus an `IsLink` property makes sense, though it was deferred to later.

Thinking out-loud: if the property exists, would it be taken into account by this method?

e.g. when `!IsLink` throw `InvalidOperationException` or `return null` immediately?

Adding this later would be a breaking change, and won't happen.

This was deferred because the semantics of `IsLink` are ambiguous. Maybe as part of this implementation it became more clear, and we can return `false` for anythin where `LinkTarget` is expected to return `null`.

cc @mklement0 @adamsitnik

Reply:
Good question, @tmds; #53577 is the follow-up issue about `IsLink`; to quote my comment from https://github.com/dotnet/runtime/issues/53577#issuecomment-853235983:

We need to get clarity on what `.IsLink` returning `true` is meant to signal:

* (a) Does it indicate _whether the file-system itself considers something a link_, as signaled on NTFS via the name-surrogate bit in the reparse-tag value?

* (b) Or does it indicate that .NET not only knows _that_ it is a link, but  _also knows how to explicitly resolve the link to its target_? That is, can members such as `.LinkTarget` or `.ResolveLinkTarget()` predictably be called?

And [later my thought](https://github.com/dotnet/runtime/issues/53577#issuecomment-854188473) was:

I'm now leaning toward (a):
* Let `.IsLink` report everything that the _file system_ thinks is a link, even if .NET doesn't understand the particular NTFS link type and cannot query its target,
* combined with with making `.LinkTarget` and `.ResolveLinkTarget()` simply return `null` for unknown link types; again, this NTFS-specific scenario could then be handled via the reparse-point API (assuming the caller understands the link format).

---

Only if (b) is implemented can the property be used to short-circuit.

And based on the implementation here it sounds non-links / unknown link types should return `null` rather than throw.

But ultimately I'm not sure whether (a) or (b) is the way to go (and perhaps the problem will rarely, if ever, arise).

Reply:
> if the property exists, would it be taken into account by this method?
e.g. when !IsLink throw InvalidOperationException or return null immediately?

I think `IsLink` will need to be consistent with `LinkTarget` i.e: `IsLink == (LinkTarget != null)` if we apply your other suggestion https://github.com/dotnet/runtime/pull/54253#discussion_r663838656 (do not query LinkTarget multiple times even if null was returned), then I think that would be enough to achieve consistency.

After that, adding `IsLink` in another release should no longer be a breaking change. Does that sounds OK to you folks?

Reply:
Really there is two common scenario: (1) manage link entity itself, (2) resolving link implicitly or explicitly in a code. Each of these scenarios requires a different approach. They are mutually exclusive. The first scenario requires a separate API like `public class SymbolicLink` or `public class MountPoint`. Second scenario requires injection in existing API/behavior that we do in the PR.
Seeing how slowly this PR is progressing, I would put it off until the next version altogether. Much more important is the _behavior_ of the existing API. For example, we already account for reparse points (RPs) in Directory.Remove. We need to implement the same in enumeration API. Then maybe we can create some internal API for RPs. This would be a great basis for further research. Looking specifically at _behavior_, I believe we probably won't need a generic API for links at all.

Reply:
@iSazonov I'm also leaning towards a.

When `IsLink` returns `false` I'd expect `LinkTarget` to throw `InvalidOperationException`. or `return null` without querying the target.

In any case, when `IsLink` returns `true`, `LinkTarget` can always still return `null` because the underlying file may have changed and no longer be a link. The user must handle this case.